File: 백준/Silver/1874. 스택 수열/스택 수열.py
import sys

n = int(sys.stdin.readline().strip())

result = []
stack = []

# 현재 가리키고 있는 값
cur = 1

for _ in range(n):
  # 수열 입력 받기
  seq = int(sys.stdin.readline().strip())
  # 현재 커서 값이 입력된 수열보다 작거나 같냐 ?
  # 그럼 커서 값을 스택에 쌓아 입력된 수열 나올 때까지
  while cur <= seq:
    stack.append(cur)
    result.append('+')
    cur += 1

  # 스택 가장 상위 값이 seq랑 같냐 ?
  # 그럼 빼서 출력해
  if stack[-1] == seq:
    stack.pop()
    result.append('-')
  # 아니라면 불가능한것이므로 NO 출력해
  else:
    result.clear()
    result.append('NO')
    break

for answer in result:
  print(answer)

File: 백준/Silver/1874. 스택 수열/스택 수열.java
import java.io.*;
import java.util.*;

// https://code.plus/course/41
// https://www.acmicpc.net/problem/1874
// 200 - 자료구조 1 -> 스택 수열
// 시간 복잡도: O(n)
public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        // n 입력
        int n = Integer.parseInt(br.readLine());
        Stack<Integer> s = new Stack<>();
        List<String> results = new ArrayList<>();
        int cur = 1;
        // n번 반복
        for(int i = 0; i < n; i++) {

            // 숫자 입력
            int num = Integer.parseInt(br.readLine());
            // 이 부분도 최대 n까지
            // 스택이 비었거나, 스택의 가장 최상단 값이 num보다 작은 경우
            while(s.empty() || s.peek() < num){
                s.push(cur++);
                results.add("+");
            }
            
            // while문을 돌고난 후
            // 스택의 최상단 값이 num과 같은 경우
            if(s.peek() == num){
                s.pop();
                results.add("-");
            // 스택의 최상단 값이 num과 같지 않은 경우
            } else{
                System.out.println("NO");
                return;
            }
        }
        results.forEach(System.out::println);
    }
}